lib/reek/configuration/configuration_file_finder.rb

```diff
@@ -17,42 +17,102 @@ class ConfigFileException < StandardError; end
-        files = dir.children.select(&:file?).sort
-        files.find { |file| file.to_s.end_with?('.reek') }
-      end
+        #
+        # Tries to find a configuration file via:
+        #   * given path (e.g. via cli switch)
+        #   * ascending down from the current directory
+        #   * looking into the home directory
+        #
+        # @return [File|nil]
+        #
+        # :reek:ControlParameter
+        def find(path: nil, current: Pathname.pwd, home: Pathname.new(Dir.home))
+          path || find_by_dir(current) || find_in_dir(home)
+        end
+
+        #
+        # Loads a configuration file from a given path.
+        # Raises on invalid data.
+        #
+        # @param path [String]
+        # @return [Hash]
+        #
+        # :reek:TooManyStatements: { max_statements: 6 }
+        def load_from_file(path)
+          return {} unless path
+          begin
+            configuration = YAML.load_file(path) || {}
+          rescue => error
+            raise ConfigFileException, "Invalid configuration file #{path}, error is #{error}"
+          end
+
+          unless configuration.is_a? Hash
+            raise ConfigFileException, "Invalid configuration file \"#{path}\" -- Not a hash"
+          end
+          configuration
+        end
+
+        private
+
+        #
+        # Recursively traverse directories down to find a configuration file.
+        #
+        # @return [File|nil]
+        #
+        def find_by_dir(start)
+          start.ascend do |dir|
+            file = find_in_dir(dir)
+            return file if file
+          end
+        end
 
-      # :reek:TooManyStatements: { max_statements: 6 }
-      def load_from_file(path)
-        return {} unless path
-        begin
-          configuration = YAML.load_file(path) || {}
-        rescue => error
-          raise ConfigFileException, "Invalid configuration file #{path}, error is #{error}"
+        #
+        # Checks a given directory for a configuration file and returns it.
+        # Raises an exception if we find more than one.
+        #
+        # @return [File|nil]
+        #
+        # :reek:FeatureEnvy
+        def find_in_dir(dir)
+          found = dir.children.select { |item| item.file? && item.to_s.end_with?('.reek') }.sort
+          if found.size > 1
+            escalate_too_many_configuration_files found, dir
+          else
+            found.first
+          end
         end
 
-        unless configuration.is_a? Hash
-          raise ConfigFileException, "Invalid configuration file \"#{path}\" -- Not a hash"
+        #
+        # Writes a proper warning message to STDERR and then exits the program.
+        #
+        # @return [undefined]
+        #
+        def escalate_too_many_configuration_files(found, directory)
+          offensive_files = found.map { |file| "'#{file.basename}'" }.join(', ')
+          warn format(TOO_MANY_CONFIGURATION_FILES_MESSAGE, offensive_files, directory)
+          exit 1
```

Comment:
But the exceptions we raise so far are different since they always happen in the scope of one single [Examiner](https://github.com/troessner/reek/blob/master/lib/reek/examiner.rb#L123). And in this scope it makes sense to let exceptions bubble up to catch them in one central place.

This is not the case here, at least I don't see it.
Put it differently: Why do you think it makes sense to throw an exception here only to let it bubble up and where would you actually handle the exception?